### PR TITLE
token selection: highlight hovered token

### PR DESCRIPTION
**Pull request by taras-yemets**
Closes https://github.com/sourcegraph/sourcegraph/issues/46077

Highlights hovered tokens with visible code-intel tooltips.

https://user-images.githubusercontent.com/25318659/210353370-df1118ee-3195-4347-bc79-5ca9ce4fd94d.mov

## Test plan
Tested manually (video attached).

## App preview:

- [Web](https://sg-web-taras-yemets-token-selection.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.00% (+0.28 kb) | 0.00% (+0.28 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/taras-yemets/token-selection-highlight-hovered-token/compare-report.html;tab=live_object) for a full comparison between the commits 6d19700 and 2e0ddc5 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Review by olafurpg — COMMENTED** [review]
I updated the CSS class to use the mark color. However, I'm still not loving this UX because 1) the color occasionally conflicts with document highlight in unpredictable ways, 2) I find it distracting to update the background color immediately as I'm moving the mouse, 3) neither IntelliJ or VS Code use this pattern.

I took a stab at highlighting only the occurrence while the tooltip is mounted, similar to what VS Code does for the hover tooltip. I got it working for mounting, but was not able to get it to unmount correctly. Neither the unmount react hooks or the CodeMirror `destroy` functions are getting called when I expect them to https://github.com/sourcegraph/sourcegraph/pull/new/olafurpg/highlight-codeintel-tooltip-occurrence

I wonder if we can solve this problem by updating the design of the popover to show some kind of pointer towards the occurrence. I noticed that the new design of the loading tooltip no longer points at the occurrence, which I think is a minor UX regression compared to the vanilla tooltips from CodeMirror. This approach has the benefit of working uniformly for mouse-driven and keyboard-driven navigation, as well as being compatible with pinned popovers.

**Comment by vovakulikov** [comment]
>I wonder if we can solve this problem by updating the design of the popover to show some kind of pointer towards the occurrence

@olafurpg are you talking about having a tail UI for the code intel tooltip in the blob view? (something like in [this demo ](https://storybook.sgdev.org/?path=/story/wildcard-popover--with-tail)


>we need to refactor the hover tooltip logic from using the CodeMirror's built-in [hoverTooltip](https://codemirror.net/docs/ref/#view.hoverTooltip) ([current implementation](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@6d19700/-/blob/client/web/src/repo/blob/codemirror/token-selection/hover.ts?L32)) method to custom relatively complex event handlers tracking the hover position and showing code-intel popovers for these positions if the conditions are me 

@taras-yemets I'm 100% for it, it may seem like a complex issue to do to solve the highlights token state problem, but I see other benefits from this refactoring like we already have tooltip position logic in the wildcard, and it would be a good improvement to have it in the blob UI as well. Also, since Popover in the wildcard is a simple react component with a controlled state, we should be possible to control its appearance and hook up to appearance triggers (to perform effects such as highlighting the target symbol) 

I don't have enough experience with CodeMirror UI to know that this would solve all our problems here, so please doubt the above!

**Comment by olafurpg** [comment]
> method to custom relatively complex event handlers tracking the hover position and showing code-intel popovers for these positions if the conditions are me

I just remembered that @fkling already implemented this logic with `hoverManager` https://sourcegraph.com/github.com/sourcegraph/sourcegraph@3a852aec3a2d9e3886ceedb85ad332340390061f/-/blob/client/web/src/repo/blob/codemirror/hovercard.tsx?L388

I used `hoverTooltip` for selection-driven navigation because it was the easiest option but the correct solution is to replace `hoverTooltip` with `hoverManager`, which would get us the desired behavior of highlighting the hovered token only while the tooltip is visible.

**Comment by olafurpg** [comment]
> since Popover in the wildcard is a simple react component with a controlled state, we should be possible to control its appearance and hook up to appearance triggers (to perform effects such as highlighting the target symbol)


@vovakulikov the CodeMirror has a lot of nice APIs to accomplish what we need. We should try to use CodeMirror idiomatically wherever possible instead of trying to move logic into React (where I personally find it sometimes more difficult to achieve custom functionality compared to the CodeMirror APIs). The CodeMirror APIs have a learning curve (facets, fields, extensions, state, decorations, ...) but once you get the hang of it and follow the conventions then a lot of impressive features start becoming easy to implement.

**Comment by vovakulikov** [comment]
@olafurpg I'm not saying that we need to move away from the codemirror API completely. I see a lot of good parts of event-driven API that code mirror has. But also, I've seen that we rely on custom logic in Monaco and codemirror tooltip position for the tooltip UI. And this bothers me a lot because it isn't so straightforward if we need to change there something (for example, add tail UI to the popover), or for example, this problem that we got with tooltip positions very recently [here](https://github.com/sourcegraph/sourcegraph/pull/45959) 

So bottom line from me here is Codemirror API is cool for the state management but UI parts from Codemirror might be non optimal for some parts of our UI (such as code intel tooltip)

**Comment by fkling** [comment]
I’m sorry that I haven’t paid closer attention to how the token selection shows hovercards.  Couldn’t it just use the `showHovercard` facet? https://sourcegraph.com/github.com/sourcegraph/sourcegraph@3a852aec3a2d9e3886ceedb85ad332340390061f/-/blob/client/web/src/repo/blob/codemirror/hovercard.tsx?L170 

The facet controls showing the tooltip/card as well as highlighting the corresponding token range.

The nice thing about CodeMirror’s API is that you’ you can easily decouple UI logic from data providers.

File: client/web/src/repo/blob/codemirror/token-selection/hover.ts [hunk 1]
```diff
@@ -222,3 +221,141 @@ const tooltipStyles = EditorView.theme({
...
+                        view.dispatch({ effects: setHoverTooltip.of(next) })
+                    })
+            )
+        }
+
+        public destroy(): void {
+            this.subscription.unsubscribe()
+        }
+    }
+)
```
Comment (taras-yemets) [inline client/web/src/repo/blob/codemirror/token-selection/hover.ts:357]: Modified version of the [non-selection-driven `hoverManager`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@3a852aec3a2d9e3886ceedb85ad332340390061f/-/blob/client/web/src/repo/blob/codemirror/hovercard.tsx?L383-541).

File: client/web/src/repo/blob/codemirror/token-selection/extension.ts [hunk 2]
```diff
@@ -100,21 +99,6 @@ export function tokenSelectionExtension(): Extension {
...
-                const atEvent = occurrenceAtMouseEvent(view, event)
-                const hoveredOccurrence = atEvent ? atEvent.occurrence : null
-                if (hoveredOccurrence !== view.state.field(hoveredOccurrenceField)) {
-                    view.dispatch({ effects: setHoveredOccurrenceEffect.of(hoveredOccurrence) })
-                    // Only pre-fetch/warmup codeintel for interactive tokens.
-                    // Users can still trigger code intel actions for
-                    // non-interactive tokens, the result just won't be
-                    // pre-fetched.
-                    if (hoveredOccurrence && isInteractiveOccurrence(hoveredOccurrence)) {
-                        warmupOccurrence(view, hoveredOccurrence)
```
Comment (taras-yemets) [inline client/web/src/repo/blob/codemirror/token-selection/extension.ts:114]: Mouse move events are now handled in the [hover manager](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@96e4d1ad55b74c0b4ccc1b7db18af23bee5b670d/-/blob/client/web/src/repo/blob/codemirror/token-selection/hover.ts?L276). 

@olafurpg, as we use a [smaller timeout](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@96e4d1ad55b74c0b4ccc1b7db18af23bee5b670d/-/blob/client/web/src/repo/blob/codemirror/token-selection/hover.ts?L279) before fetching the code-intel data now (25ms compared to the [initial 200ms](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@35ecf2f7ca68d54bb17376c92cece67375d7acfa/-/blob/client/web/src/repo/blob/codemirror/token-selection/hover.ts?L31:77&popover=pinned)), do we still need to warm up the occurrence on hover?

File: client/web/src/repo/blob/codemirror/token-selection/hover.ts [hunk 3]
```diff
@@ -222,3 +221,141 @@ const tooltipStyles = EditorView.theme({
...
+                                tap(range => {
+                                    const currentRange = view.state.field(hoverTooltip)?.range
+                                    if (currentRange && !isEqual(currentRange, range)) {
+                                        // different occurrence is hovered - hide the tooltip for the previous one and fetch the new one
+                                        view.dispatch({ effects: setHoverTooltip.of(null) })
+                                    }
+                                }),
+                                switchMap(range =>
+                                    from(getHoverTooltip(view, range.from)).pipe(
+                                        catchError(() => of(null)),
```
Comment (taras-yemets) [inline client/web/src/repo/blob/codemirror/token-selection/hover.ts:340]: Needed to catch errors mentioned in
- https://github.com/sourcegraph/sourcegraph/issues/46071
(a [similar approach](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@96e4d1ad55b74c0b4ccc1b7db18af23bee5b670d/-/blob/client/web/src/repo/blob/codemirror/token-selection/keybindings.ts?L32) is used in the selection-driven nav 'Space' handler).

**Comment by olafurpg** [comment]
The new behavior feels great! I tested it quickly locally and noticed that the token is not highlighted when using the pinned popover https://sourcegraph.test:3443/github.com/scalameta/metals/-/blob/tests/unit/src/test/scala/tests/IdentifierComparatorSuite.scala?L16:3&popover=pinned#tab=references

CleanShot 2023-01-04 at 15 01 41@2x


Will review the code in more detail later

File: client/web/src/repo/blob/codemirror/token-selection/extension.ts (hunk 2 above)
Comment (olafurpg) [reply client/web/src/repo/blob/codemirror/token-selection/extension.ts:114]: I leave it to you to decide if we need a warmup or not. Waiting 25ms seems fine

File: client/web/src/repo/blob/codemirror/token-selection/hover.ts (hunk 3 above)
Comment (olafurpg) [reply client/web/src/repo/blob/codemirror/token-selection/hover.ts:340]: We should update the API to return `null | undefined`

File: client/web/src/repo/blob/codemirror/token-selection/hover.ts [hunk 4]
```diff
@@ -222,3 +221,141 @@ const tooltipStyles = EditorView.theme({
...
+                            }
+
+                            const currentTooltip = view.state.field(hoverTooltip)
+                            if (!currentTooltip) {
+                                return true
+                            }
+
+                            return !isOffsetInHoverRange(offset, currentTooltip.range)
+                        }),
+
```
Comment (olafurpg) [inline client/web/src/repo/blob/codemirror/token-selection/hover.ts:322]: The original `hoverManager` has a `scan` step to mirror the behavior of the hover tooltip in VS Code wrt. to the direction movement of the mouse. Was there a specific reason you left this out of the re-implementation?

https://sourcegraph.com/github.com/sourcegraph/sourcegraph@taras-yemets/token-selection-highlight-hovered-token/-/blob/client/web/src/repo/blob/codemirror/hovercard.tsx?L439

**Review by olafurpg — CHANGES_REQUESTED** [review]
Two blocking comments: `scan` and pinned popovers. Otherwise, this looks great!

File: client/web/src/repo/blob/codemirror/token-selection/hover.ts (hunk 4 above)
Comment (taras-yemets) [reply client/web/src/repo/blob/codemirror/token-selection/hover.ts:322]: > Was there a specific reason you left this out of the re-implementation?

It seemed to work fine without taking the cursor movement direction into consideration.
But I didn't it on long tokens, for example 🤔 When the hovered token is wider than the code-intel tooltip, hovering over the end of the token, seeing the code-intel popup, and then trying to hover over it may be impossible (the tooltip will hide after the cursor leaves the token).
Will double-check. Thanks for flagging it!

Comment (taras-yemets) [reply client/web/src/repo/blob/codemirror/token-selection/hover.ts:344]: Addressed in https://github.com/sourcegraph/sourcegraph/pull/45951/commits/fef6f59a1e73fc8ce0322ba5017d1a3c33f47228

File: client/shared/src/codeintel/api.ts [hunk 5]
```diff
@@ -34,7 +34,7 @@ interface CodeIntelAPI {
         context: sourcegraph.ReferenceContext
     ): Observable<clientType.Location[]>
     getImplementations(parameters: TextDocumentPositionParameters): Observable<clientType.Location[]>
-    getHover(textParameters: TextDocumentPositionParameters): Observable<HoverMerged | null>
+    getHover(textParameters: TextDocumentPositionParameters): Observable<HoverMerged | null | undefined>
```
Comment (taras-yemets) [inline client/shared/src/codeintel/api.ts:37]: As suggested in https://github.com/sourcegraph/sourcegraph/pull/45951#discussion_r1061606780 add `undefined` to `getHover` return type (see [`ProviderResult` type definition](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@01c578d69914fa4980a8c7e39363e7d928991d27/-/blob/client/shared/src/codeintel/legacy-extensions/api.ts?L442:1-452:37)).

**Comment by taras-yemets** [comment]
Thank you for your review, @olafurpg!
Since your last review:
- added `scan` to the hover manager to respect the mouse cursor movement direction (https://github.com/sourcegraph/sourcegraph/pull/45951/commits/fef6f59a1e73fc8ce0322ba5017d1a3c33f47228)
- highlight token with pinned popover

https://user-images.githubusercontent.com/25318659/211010169-1d198fe7-417c-439d-8d2b-c1dfb9b0a27b.mov

**Review by philipp-spiess — APPROVED** [review]
This looks great :)

File: client/web/src/repo/blob/codemirror/token-selection/hover.ts [hunk 6]
```diff
@@ -222,3 +262,177 @@ const tooltipStyles = EditorView.theme({
...
+                                return true
+                            }
+
+                            const currentTooltip = view.state.field(hoverTooltip)
+                            if (!currentTooltip) {
+                                return true
+                            }
+
+                            return !isOffsetInHoverRange(offset, currentTooltip.range)
+                        }),
```
[the remark this answers is not in the document]
Comment (olafurpg) [reply client/web/src/repo/blob/codemirror/token-selection/hover.ts:343]: The duplication shouldn't be a problem because the code paths are never used at the same time. If we stick with selection-driven navigation then we should be able remove the duplicate code in hovercard.tsx

[the remark this answers is not in the document]
Comment (philipp-spiess) [reply client/web/src/repo/blob/codemirror/token-selection/hover.ts:343]: OH gotcha I haven't realized those are mutually exclusive.
